crates/key-value/src/util.rs

```diff
@@ -31,3 +39,167 @@ impl StoreManager for DelegatingStoreManager {
+pub struct CachingStoreManager<T> {
+    inner: T,
+}
+
+impl<T> CachingStoreManager<T> {
+    pub fn new(inner: T) -> Self {
+        Self { inner }
+    }
+}
+
+#[async_trait]
+impl<T: StoreManager> StoreManager for CachingStoreManager<T> {
+    async fn get(&self, name: &str) -> Result<Arc<dyn Store>, Error> {
+        Ok(Arc::new(CachingStore {
+            inner: self.inner.get(name).await?,
+            state: AsyncMutex::new(CachingStoreState::default()),
+        }))
+    }
+}
+
+#[derive(Default)]
+struct CachingStoreState {
+    cache: HashMap<String, Option<Vec<u8>>>,
+    previous_task: Option<JoinHandle<Result<(), Error>>>,
+}
+
+impl CachingStoreState {
+    /// Wrap the specified task in an outer task which waits for `self.previous_task` before proceeding, and spawn
+    /// the result.  This ensures that write order is preserved.
+    fn spawn(&mut self, task: impl Future<Output = Result<(), Error>> + Send + 'static) {
+        let previous_task = self.previous_task.take();
+        self.previous_task = Some(task::spawn(async move {
+            if let Some(previous_task) = previous_task {
+                previous_task
+                    .await
+                    .map_err(|e| Error::Io(format!("{e:?}")))??
+            }
+
+            task.await
+        }))
+    }
+}
+
+struct CachingStore {
+    inner: Arc<dyn Store>,
+    state: AsyncMutex<CachingStoreState>,
+}
+
+#[async_trait]
+impl Store for CachingStore {
+    async fn get(&self, key: &str) -> Result<Vec<u8>, Error> {
+        // Retrieve the specified value from the cache, lazily populating the cache as necessary.
+
+        match self.state.lock().await.cache.entry(key.to_owned()) {
+            Entry::Vacant(e) => {
+                let value = match self.inner.get(key).await {
+                    Ok(value) => Some(value),
+                    Err(Error::NoSuchKey) => None,
+                    e => return e,
+                };
+
+                e.insert(value.clone());
+
+                value
+            }
+            Entry::Occupied(e) => e.get().clone(),
+        }
+        .ok_or(Error::NoSuchKey)
+    }
+
+    async fn set(&self, key: &str, value: &[u8]) -> Result<(), Error> {
+        // Update the cache and spawn a task to update the backing store asynchronously.
+
+        let mut state = self.state.lock().await;
+
+        state.cache.insert(key.to_owned(), Some(value.to_owned()));
+
+        let inner = self.inner.clone();
+        let key = key.to_owned();
+        let value = value.to_owned();
+        state.spawn(async move { inner.set(&key, &value).await });
+
+        Ok(())
+    }
+
+    async fn delete(&self, key: &str) -> Result<(), Error> {
+        // Update the cache and spawn a task to update the backing store asynchronously.
+
+        let mut state = self.state.lock().await;
+
+        state.cache.insert(key.to_owned(), None);
```

Comment:
Why is this being set to `None` instead of being removed from the cache? I see that we are checking for `None` values for `get` and `get_keys` so implementation is sound, but i curious

Reply:
If we remove it from the cache entirely, and then the guest asks for that key later, it won't be in the cache, meaning we'll need to hit the backing store (e.g. SQLite) to find out if there's a value or not.  By storing `None`, we know the value's been deleted without needing to consult the backing store.  This is sometimes called a "tombstone" by the way.